[19.09] Structured access to Galaxy internals for (Interactive) Tools

jmchilton opened the pull request.

jmchilton committed:
Structured access to Galaxy internals for ITs.

jmchilton committed:
Allow per-destination overrride of galaxy_url for tools.

The proper interaction between tools and jobs for determining Galaxy's URL from the destination of the tool.

bgruening commented on tools/interactive/interactivetool_jupyter_notebook.xml:
```diff
@@ -9,18 +9,11 @@
     <environment_variables>
-        <environment_variable name="HISTORY_ID" strip="True">${__app__.security.encode_id($jupyter_notebook.history_id)}</environment_variable> <!-- FIXME: Warning: The use of __app__ is deprecated and will break backward compatibility in the near future -->
-        <environment_variable name="REMOTE_HOST">${__app__.config.galaxy_infrastructure_url}</environment_variable> <!-- FIXME: Warning: The use of __app__ is deprecated and will break backward compatibility in the near future -->
+        <environment_variable name="HISTORY_ID">$__history_id</environment_variable>
```
@jmchilton can we make this consistent with `$__tool_directory__` and end with `__` as well?

bgruening commented on tools/interactive/interactivetool_jupyter_notebook.xml:
```diff
@@ -9,18 +9,11 @@
-        <environment_variable name="REMOTE_HOST">${__app__.config.galaxy_infrastructure_url}</environment_variable> <!-- FIXME: Warning: The use of __app__ is deprecated and will break backward compatibility in the near future -->
+        <environment_variable name="HISTORY_ID">$__history_id</environment_variable>
+        <environment_variable name="REMOTE_HOST">$__galaxy_url</environment_variable>
         <environment_variable name="GALAXY_WEB_PORT">8080</environment_variable>
```
what about a `galaxy_infrastructure_web_port`?

jmchilton replied:
That was removed in dev I think (with https://github.com/galaxyproject/galaxy/commit/e50807a7c6ce050cedab623546f43d1b2b75ac52) - seems like we decided it doesn't do anything most of the time? Is GALAXY_WEB_PORT actually used by the container?

bgruening replied:
No not yet. You are correct, let's leave this out as long as we have no use-case for it.

jmchilton committed:
Cleanup structured IT internals access commit.

- Unit test fixes and added tests.
- Change __history_id to __history_id__ per comment from @bgruening

jmchilton committed:
Update galaxy.xsd for injecting API keys.

jmchilton committed:
Fix id convention for test interactive tools.

jmchilton committed:
Separate new inject tests into new tool.

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -543,9 +545,16 @@ def __build_environment_variables(self):
+            inject = environment_variable_def.get("inject")
+            if inject == "api_key":
+                from galaxy.managers import api_keys
+                environment_variable_template = api_keys.ApiKeyManager(self.app).get_or_create_api_key(self._user)
```
self._user would be `None` for anonymous users, failing with:
```
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/runners/__init__.py", line 230, in prepare_job
    job_wrapper.prepare()
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/__init__.py", line 1066, in prepare
    self.command_line, self.extra_filenames, self.environment_variables = tool_evaluator.build()
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/evaluation.py", line 484, in build
    raise e
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/evaluation.py", line 481, in build
    self.__build_environment_variables()
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/evaluation.py", line 551, in __build_environment_variables
    environment_variable_template = api_keys.ApiKeyManager(self.app).get_or_create_api_key(self._user)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/managers/api_keys.py", line 22, in get_or_create_api_key
    if user.api_keys:
AttributeError: 'NoneType' object has no attribute 'api_keys'
```
Maybe we can raise a more explicit exception in the `_user` property ?

jmchilton replied:
Maybe I'll just set the variable to empty - if the tool author wants to require a user they should use ``required_login`` so presumably if an anonymous user is using this tool the API key is optional?

cc @bgruening

bgruening replied:
Agreed, this seems logical. Thanks!

jmchilton committed:
Allow anonymous users to use tools with injected API keys.

https://github.com/galaxyproject/galaxy/pull/8897/files#r342997635

bgruening reviewed (approved):
LGTM, there is one failing test.
Thanks @jmchilton

mvdbeek commented:
```
galaxy.queue_worker: ERROR: Error queueing async task: '{'task': 'recalculate_user_disk_usage', 'kwargs': {}}'. for control.main@42e141c29f5c
Traceback (most recent call last):
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: kombu_queue.name

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/galaxy/lib/galaxy/queue_worker.py", line 128, in send_task
    with Consumer(self.connection, on_message=self.on_response, queues=callback_queue, no_ack=True):
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/messaging.py", line 386, in __init__
    self.revive(self.channel)
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/messaging.py", line 408, in revive
    self.declare()
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/messaging.py", line 421, in declare
    queue.declare()
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/entity.py", line 609, in declare
    self._create_queue(nowait=nowait, channel=channel)
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/entity.py", line 618, in _create_queue
    self.queue_declare(nowait=nowait, passive=False, channel=channel)
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/entity.py", line 653, in queue_declare
    nowait=nowait,
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/transport/virtual/base.py", line 531, in queue_declare
    self._new_queue(queue, **kwargs)
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/transport/sqlalchemy/__init__.py", line 82, in _new_queue
    self._get_or_create(queue)
  File "/galaxy_venv3/lib/python3.5/site-packages/kombu/transport/sqlalchemy/__init__.py", line 76, in _get_or_create
    self.session.commit()
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 1027, in commit
    self.transaction.commit()
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 494, in commit
    self._prepare_impl()
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 473, in _prepare_impl
    self.session.flush()
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 2459, in flush
    self._flush(objects)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 2597, in _flush
    transaction.rollback(_capture_exception=True)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 153, in reraise
    raise value
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 2557, in _flush
    flush_context.execute()
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
    rec.execute(self)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/unitofwork.py", line 589, in execute
    uow,
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/persistence.py", line 245, in save_obj
    insert,
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/persistence.py", line 1138, in _emit_insert_statements
    statement, params
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1248, in _execute_context
    e, statement, parameters, cursor, context
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1466, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: kombu_queue.name
[SQL: INSERT INTO kombu_queue (name) VALUES (?)]
[parameters: ('57a9745c-dedf-4ede-b1eb-781711f580e9',)]
(Background on this error at: http://sqlalche.me/e/gkpj)
```
I don't think the failing test is related.

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -543,9 +545,19 @@ def __build_environment_variables(self):
+                    from galaxy.managers import api_keys
+                    environment_variable_template = api_keys.ApiKeyManager(self.app).get_or_create_api_key(self._user)
+                else:
+                    environment_variable_template = ""
```
Can we raise an Exception here ? They will be displayed in the dataset misc blurb thingy. If the tool asked for the api key and we can't provide it we should probably error with a useful message.

jmchilton replied:
Discussed this here https://github.com/galaxyproject/galaxy/pull/8897#discussion_r343050378. I think if the tool doesn't specify ``require_login="true"`` then the API key should be assumed to be optional.

mvdbeek replied:
makes sense
